File: backend/internal/models/storage_pool.go
package models

import (
	"time"
	"gorm.io/gorm"
)

// StoragePool 存储池模型
type StoragePool struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name        string         `gorm:"uniqueIndex;not null" json:"name"`
	Type        string         `json:"type"`          // mergerfs, btrfs, zfs, lvm
	Status      string         `json:"status"`        // active, inactive, error, degraded
	MountPoint  string         `gorm:"not null" json:"mountPoint"`
	TotalSize   uint64         `json:"totalSize"`
	UsedSize    uint64         `json:"usedSize"`
	FreeSize    uint64         `json:"freeSize"`
	Description string         `json:"description"`
	Config      string         `gorm:"type:json" json:"config"`
	CreatedBy   string         `json:"createdBy"`

	PoolDisks  []PoolDisk     `gorm:"foreignKey:PoolID" json:"poolDisks,omitempty"`
	Snapshots  []PoolSnapshot `gorm:"foreignKey:PoolID" json:"snapshots,omitempty"`
}

// PoolDisk 存储池磁盘关联
type PoolDisk struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	PoolID     uint   `gorm:"not null" json:"poolId"`
	Device     string `gorm:"not null" json:"device"`
	Size       uint64 `json:"size"`
	Status     string `json:"status"`     // active, failed, removed
	Priority   int    `json:"priority"`    // mergerfs 优先级
	BranchPath string `json:"branchPath"` // 挂载路径
}

// PoolSnapshot 存储池快照
type PoolSnapshot struct {
	ID         uint           `gorm:"primarykey" json:"id"`
	CreatedAt  time.Time      `json:"createdAt"`
	UpdatedAt  time.Time      `json:"updatedAt"`
	DeletedAt  gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	PoolID      uint       `gorm:"not null" json:"poolId"`
	Name        string     `gorm:"not null" json:"name"`
	Description string     `json:"description"`
	Size        uint64     `json:"size"`
	Status      string     `json:"status"`    // creating, completed, deleting, error
	CompletedAt *time.Time `json:"completedAt,omitempty"`
}

// RAIDConfig RAID配置
type RAIDConfig struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name       string `json:"name"`
	Level      string `json:"level"`        // 0, 1, 5, 6, 10, 1+0, etc.
	Devices    string `gorm:"type:json" json:"devices"`  // JSON数组
	Status     string `json:"status"`       // active, degraded, failed, recovering
	Size       uint64 `json:"size"`
	UUID       string `json:"uuid"`
	MountPoint string `json:"mountPoint"`
}

// PhysicalVolume 物理卷
type PhysicalVolume struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Device string `json:"device"`
	VGName string `json:"vgName"`
	Size   uint64 `json:"size"`
	Free   uint64 `json:"free"`
	UUID   string `json:"uuid"`
	Status string `json:"status"` // active, missing, exported
}

// VolumeGroup 卷组
type VolumeGroup struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name    string `json:"name"`
	Size    uint64 `json:"size"`
	Free    uint64 `json:"free"`
	PVCount int    `json:"pvCount"`
	LVCount int    `json:"lvCount"`
	UUID    string `json:"uuid"`
	Status  string `json:"status"` // active, partial, exported
}

// LogicalVolume 逻辑卷
type LogicalVolume struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name       string `json:"name"`
	VGName     string `json:"vgName"`
	Size       uint64 `json:"size"`
	Path       string `json:"path"`
	UUID       string `json:"uuid"`
	MountPoint string `json:"mountPoint"`
	Status     string `json:"status"` // active, inactive, snapshot merge
}

File: frontend/src/types/storage_pool.ts
// Storage Pool Type Definitions

export interface StoragePool {
  id: number
  name: string
  type: 'mergerfs' | 'btrfs' | 'zfs' | 'lvm'
  status: 'active' | 'inactive' | 'creating' | 'deleting' | 'error' | 'degraded'
  mountPoint: string
  totalSize: number
  usedSize: number
  freeSize: number
  description: string
  config: MergerFSConfig | null
  poolDisks: PoolDisk[]
  snapshots: PoolSnapshot[]
  createdAt: string
  updatedAt: string
}

export interface MergerFSConfig {
  branches: BranchConfig[]
  category: 'create' | 'mv' | 'epall' | 'epff' | 'epmfs' | 'eplus' | 'eprand' | 'lus' | 'lus' | 'mfs'
  minfreespace: string
  direct_io: boolean
  async_read: boolean
  use_ino: boolean
  hard_remove: boolean
  auto_unshare: boolean
  follow_symlinks: boolean
  link_exas: boolean
}

export interface BranchConfig {
  path: string
  mode: 'ro' | 'rw'
  priority: number
}

export interface PoolDisk {
  id?: number
  poolId?: number
  device: string
  size: number
  status?: 'active' | 'failed' | 'removed' | 'ro' | 'rw'
  priority: number
  branchPath: string
  used?: number
  free?: number
  createdAt?: string
  updatedAt?: string
}

export interface PoolSnapshot {
  id: number
  poolId: number
  name: string
  description: string
  size: number
  status: 'creating' | 'completed' | 'deleting' | 'error'
  createdAt: string
  completedAt: string | null
}

export interface StoragePoolCreateRequest {
  name: string
  type: 'mergerfs' | 'btrfs' | 'zfs' | 'lvm'
  mountPoint: string
  description?: string
  config?: MergerFSConfig
  disks: PoolDisk[]
}

export interface StoragePoolUpdateRequest {
  description?: string
  config?: MergerFSConfig
}

export interface PoolStatusResponse {
  name: string
  status: string
  branches: BranchInfo[]
  totalSize: number
  usedSize: number
  freeSize: number
}

export interface BranchInfo {
  path: string
  mode: 'ro' | 'rw'
  priority: number
  size: number
  used: number
  free: number
}

export interface MergerFSCategory {
  value: string
  label: string
  description: string
}

export const MERGERFS_CATEGORIES: MergerFSCategory[] = [
  { value: 'create', label: 'Create', description: 'Create on first branch with space' },
  { value: 'mv', label: 'Move', description: 'Move to branch with most space' },
  { value: 'epall', label: 'EP All', description: 'Round-robin across all branches' },
  { value: 'epff', label: 'EP First', description: 'Place on first available branch' },
  { value: 'epmfs', label: 'EP MFS', description: 'Place on branch with most free space' },
  { value: 'eplus', label: 'EP Plus', description: 'Place on branch with most plus space' },
  { value: 'eprand', label: 'EP Random', description: 'Place on random branch' },
  { value: 'lus', label: 'LUS', description: 'Least used space' },
  { value: 'mfs', label: 'MFS', description: 'Most free space' }
]

export const STORAGE_POOL_STATUS_COLORS = {
  active: 'green',
  inactive: 'gray',
  creating: 'blue',
  deleting: 'orange',
  error: 'red',
  degraded: 'yellow'
}

export const STORAGE_POOL_TYPE_LABELS = {
  mergerfs: 'MergerFS',
  btrfs: 'Btrfs',
  zfs: 'ZFS',
  lvm: 'LVM'
}